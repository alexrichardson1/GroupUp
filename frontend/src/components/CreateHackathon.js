import React, { Component } from "react";
import axios from "axios";
import "components/styles.css";
import { config } from "Constants";
import { Form, Button, Container, Col } from "react-bootstrap";
import { withRouter } from "react-router-dom";
import NavBar from "components/NavBar";

async function addProject(data) {
  var result = {};
  await axios
    .post(`${config.API_URL}/project/add`, data)
    .then((res) => {
      const group = res.data;
      result = group;
    })
    .catch((error) => {
      console.log(error);
    });
  return result;
}

class CreateHackathon extends Component {
  constructor(props) {
    super(props);
    this.state = {
      hackathonName: "",
      requirements: [],
      description: "",
      hours: "",
      date: "",
      location: "",
    };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  async componentDidMount() {}

  handleInputChange(event) {
    const target = event.target;
    const value = target.type === "checkbox" ? target.checked : target.value;

    this.setState({
      [target.name]: value,
    });
  }

  handleSubmit = async (e) => {
    e.preventDefault();
    const info = {
      name: this.state.hackathonName,
      requirements: this.state.requirements.split(", "),
      description: this.state.description,
      hours: this.state.hours,
      date: new Date(this.state.date).toISOString(),
      location: this.state.location,
    };
    await addProject(info);
    window.location.reload();
  };

  genFormComponent = (label, placeholder, stateName, mutedText) => {
    return (
      <Form.Group
        className="formBox"
        as={Col}
        controlId="formLeaderName"
        key="formLeaderName"
      >
        <Form.Label>{label}</Form.Label>
        <Form.Control
          type="text"
          placeholder={placeholder}
          name={stateName}
          value={this.state[stateName]}
          onChange={this.handleInputChange}
        />
        <Form.Text className="text-muted">{mutedText}</Form.Text>
      </Form.Group>
    );
  };

  render() {
    return (
      <Container>
        <NavBar
          renderBool={[true, true, true, false]}
          create={true}
          id={this.props.id}
        />
        <h1>Adding your Hackathon to our list</h1>
        <Form onSubmit={this.handleSubmit}>
          {this.genFormComponent(
            "Hackathon Name",
            "e.g. University of Paris Hackathon 2021",
            "hackathonName",
            ""
          )}

          {this.genFormComponent(
            "Filter Requirements",
            "e.g. Programming Language, Timezone [if remote], Version Control",
            "requirements",
            'These are the compulsory requirements that groups advertising for your hackathon must specify. For example, you might want groups to specify "Timezone" when advertising, if your Hackathon is remote.'
          )}

          {this.genFormComponent(
            "Description",
            "e.g. The Annual Hackathon event for University of Paris, where you will be creating a web-based game!",
            "description",
            "Include a little bit about your hackathon, to make sure users are on the right page."
          )}

          {this.genFormComponent(
            "Hours",
            "24",
            "hours",
            "The length of your hackathon as number of hours."
          )}

          <Form.Group
            className="formBox"
            as={Col}
            controlId="formLeaderName"
            key="formLeaderName"
          >
            <Form.Label>Start Date</Form.Label>
            <Form.Control
              type="date"
              placeholder="dfsd"
              name="date"
              value={this.state.date}
              onChange={this.handleInputChange}
            />
          </Form.Group>

          {this.genFormComponent(
            "Location",
            "e.g. Paris, France",
            "location",
            ""
          )}

          <Button variant="primary" type="submit">
            Submit
          </Button>
        </Form>
      </Container>
    );
  }
}

export default withRouter(CreateHackathon);
