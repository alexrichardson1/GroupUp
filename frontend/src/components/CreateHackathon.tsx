import { Component } from "react";
import "components/styles.css";
import { Form, Button, Container, Col, Alert, Spinner } from "react-bootstrap";
import { withRouter } from "react-router-dom";
import NavBar from "components/NavBar";
import { Redirect } from "react-router";
import { ProjectT } from "types/types";
import { addProject } from "common/api";
import { allFieldsNotFilledIn } from "common/render";

interface Props {
  id: number;
}

interface State {
  hackathonName: string;
  requirements: string;
  description: string;
  hours: string;
  date: string;
  location: string;
  invalid: boolean;
  invalid2: boolean;
  redirect: boolean;
}

class CreateHackathon extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      hackathonName: "",
      requirements: "",
      description: "",
      hours: "",
      date: "",
      location: "",
      invalid: false,
      invalid2: false,
      redirect: false,
    };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  override async componentDidMount() {}

  handleInputChange(event: any) {
    const target = event.target;
    const value = target.type === "checkbox" ? target.checked : target.value;

    //@ts-ignore
    this.setState({
      [target.name]: value,
    });
  }

  handleSubmit = async (e: any) => {
    e.preventDefault();
    if (
      this.state.hackathonName === "" ||
      this.state.requirements.length === 0 ||
      this.state.description === "" ||
      this.state.hours === "" ||
      isNaN(parseInt(this.state.hours)) ||
      this.state.date === "" ||
      this.state.location === ""
    ) {
      this.setState({ invalid: true });
    } else {
      const info: ProjectT = {
        name: this.state.hackathonName,
        requirements: this.state.requirements.split(", "),
        description: this.state.description,
        hours: parseInt(this.state.hours),
        date: new Date(this.state.date).toISOString(),
        location: this.state.location,
      };
      await addProject(info);
      this.setState({ invalid: false });
      const TIMEOUT = 3000;
      //@ts-ignore
      this.id = setTimeout(() => this.setState({ redirect: true }), TIMEOUT);
      this.setState({ invalid2: true });
    }
  };

  genFormComponent = (
    label: string,
    placeholder: string,
    stateName: string,
    mutedText: string
  ) => {
    return (
      <Form.Group
        className="formBox"
        as={Col}
        controlId="formLeaderName"
        key="formLeaderName"
      >
        <Form.Label>{label}</Form.Label>
        <Form.Control
          type="text"
          placeholder={placeholder}
          name={stateName}
          //@ts-ignore
          value={this.state[stateName]}
          onChange={this.handleInputChange}
          //@ts-ignore
          isInvalid={this.state[stateName].length === 0}
        />
        <Form.Text className="text-muted">{mutedText}</Form.Text>
      </Form.Group>
    );
  };

  hackathonNowAvailable() {
    if (this.state.invalid2) {
      return <Alert variant="success">Hackathon is now available!</Alert>;
    }
    return null;
  }

  override render() {
    return this.state.redirect ? (
      <Redirect to="/home" />
    ) : (
      <Container>
        <NavBar
          renderBool={[true, true, true, false]}
          create={true}
          hackathon={true}
          id={this.props.id}
        />
        <h1>Adding your Hackathon to our list</h1>
        {allFieldsNotFilledIn(this.state.invalid)}
        {this.hackathonNowAvailable()}
        {!this.state.invalid2 ? (
          <Form onSubmit={this.handleSubmit}>
            {this.genFormComponent(
              "Hackathon Name",
              "e.g. University of Paris Hackathon 2021",
              "hackathonName",
              ""
            )}

            {this.genFormComponent(
              "Filter Requirements",
              "e.g. Programming Language, Timezone [if remote], Version Control",
              "requirements",
              'These are the compulsory requirements that groups advertising for your hackathon must specify. For example, you might want groups to specify "Timezone" when advertising, if your Hackathon is remote.'
            )}

            {this.genFormComponent(
              "Description",
              "e.g. The Annual Hackathon event for University of Paris, where you will be creating a web-based game!",
              "description",
              "Include a little bit about your hackathon, to make sure users are on the right page."
            )}

            <Form.Group
              className="formBox"
              as={Col}
              controlId="formLeaderName"
              key="formLeaderName"
            >
              <Form.Label>Length of Hackathon (Hours)</Form.Label>
              <Form.Control
                type="number"
                placeholder="24"
                name="hours"
                value={this.state.hours}
                onChange={this.handleInputChange}
                isInvalid={this.state.hours.length === 0}
              />
            </Form.Group>

            <Form.Group
              className="formBox"
              as={Col}
              controlId="formLeaderName"
              key="formLeaderName"
            >
              <Form.Label>Start Date</Form.Label>
              <Form.Control
                type="date"
                placeholder="dfsd"
                name="date"
                value={this.state.date}
                onChange={this.handleInputChange}
                isInvalid={this.state.date.length === 0}
              />
            </Form.Group>

            {this.genFormComponent(
              "Location",
              "e.g. Paris, France",
              "location",
              ""
            )}

            <Button variant="primary" type="submit">
              Submit
            </Button>
          </Form>
        ) : (
          <Spinner animation="grow" />
        )}
      </Container>
    );
  }
}

//@ts-ignore
export default withRouter(CreateHackathon);
