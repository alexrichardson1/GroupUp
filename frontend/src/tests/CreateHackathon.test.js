import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import CreateHackathon from "components/CreateHackathon";

test("renders correctly", () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <CreateHackathon />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
