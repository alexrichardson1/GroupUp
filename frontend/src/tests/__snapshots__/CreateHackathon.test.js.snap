// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container"
>
  <nav
    className="navBar navbar navbar-expand-lg navbar-light fixed-top"
  >
    <span
      className="navbar-brand"
      style={
        Object {
          "color": "white",
        }
      }
    >
      <img
        alt="group up logo"
        className="logoImage"
        href="/home"
        onClick={[Function]}
        src="https://i.imgur.com/w92t1xH.png"
      />
    </span>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      className="navbar-toggler collapsed"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
    >
      <div
        className="mr=auto navbar-nav"
        onKeyDown={[Function]}
      >
        <span
          className="navbar-brand"
          style={
            Object {
              "color": "rgb(238, 237, 237)",
            }
          }
        >
           
        </span>
        <a
          className="nav-link"
          data-rb-event-key="/home"
          href="/home"
          onClick={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "color": "rgb(238, 237, 237)",
            }
          }
        >
          Home
          <img
            alt="drop down arrow"
            className="processImage"
            src="/right-caret.svg"
          />
        </a>
        <a
          className="nav-link"
          data-rb-event-key="/selection"
          href="/selection"
          onClick={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "color": "rgb(238, 237, 237)",
            }
          }
        >
          Select Hackathon
          <img
            alt="drop down arrow"
            className="processImage"
            src="/right-caret.svg"
          />
        </a>
        <span
          className="navbar-brand"
          style={
            Object {
              "color": "rgb(238, 237, 237)",
            }
          }
        >
          Register a hackathon
        </span>
      </div>
      <div
        className="ms-auto navbar-nav"
        onKeyDown={[Function]}
      >
        <a
          className="ml-auto navHelp navbar-brand"
          href="/help"
          id="collasible-nav-dropdown"
          onClick={[Function]}
          title="Help"
        >
          <img
            alt="help"
            className="splitterImage"
            src="/info-circle.svg"
          />
        </a>
      </div>
    </div>
  </nav>
  <h1>
    Adding your Hackathon to our list
  </h1>
  <form
    className=""
    onSubmit={[Function]}
  >
    <div
      className="formBox form-group col"
    >
      <label
        className="form-label"
        htmlFor="formLeaderName"
      >
        Hackathon Name
      </label>
      <input
        className="form-control is-invalid"
        id="formLeaderName"
        name="hackathonName"
        onChange={[Function]}
        placeholder="e.g. University of Paris Hackathon 2021"
        type="text"
        value=""
      />
      <small
        className="text-muted form-text"
      >
        
      </small>
    </div>
    <div
      className="formBox form-group col"
    >
      <label
        className="form-label"
        htmlFor="formLeaderName"
      >
        Filter Requirements
      </label>
      <input
        className="form-control is-invalid"
        id="formLeaderName"
        name="requirements"
        onChange={[Function]}
        placeholder="e.g. Programming Language, Timezone [if remote], Version Control"
        type="text"
        value=""
      />
      <small
        className="text-muted form-text"
      >
        These are the compulsory requirements that groups advertising for your hackathon must specify. For example, you might want groups to specify "Timezone" when advertising, if your Hackathon is remote.
      </small>
    </div>
    <div
      className="formBox form-group col"
    >
      <label
        className="form-label"
        htmlFor="formLeaderName"
      >
        Description
      </label>
      <input
        className="form-control is-invalid"
        id="formLeaderName"
        name="description"
        onChange={[Function]}
        placeholder="e.g. The Annual Hackathon event for University of Paris, where you will be creating a web-based game!"
        type="text"
        value=""
      />
      <small
        className="text-muted form-text"
      >
        Include a little bit about your hackathon, to make sure users are on the right page.
      </small>
    </div>
    <div
      className="formBox form-group col"
    >
      <label
        className="form-label"
        htmlFor="formLeaderName"
      >
        Length of Hackathon (Hours)
      </label>
      <input
        className="form-control is-invalid"
        id="formLeaderName"
        name="hours"
        onChange={[Function]}
        placeholder="24"
        type="number"
        value=""
      />
    </div>
    <div
      className="formBox form-group col"
    >
      <label
        className="form-label"
        htmlFor="formLeaderName"
      >
        Start Date
      </label>
      <input
        className="form-control is-invalid"
        id="formLeaderName"
        name="date"
        onChange={[Function]}
        placeholder="dfsd"
        type="date"
        value=""
      />
    </div>
    <div
      className="formBox form-group col"
    >
      <label
        className="form-label"
        htmlFor="formLeaderName"
      >
        Location
      </label>
      <input
        className="form-control is-invalid"
        id="formLeaderName"
        name="location"
        onChange={[Function]}
        placeholder="e.g. Paris, France"
        type="text"
        value=""
      />
      <small
        className="text-muted form-text"
      >
        
      </small>
    </div>
    <button
      className="btn btn-primary"
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
